### PR TITLE
WIP postcss-selector-parser

### PR DESCRIPTION
**Pull request by devinus**

### CODE REVIEW

**Comment by devinus** [comment]
@motdotla @elbuo8 This replaces our hacky regex based solution with a real CSS selector parser.

Tested against several documents, there are no differences in output. It's unfortunately 1.1x-1.2x slower depending on input, but I hope to make up for that if we move to PostHTML. Initial benchmarks using the `posthtml` branch puts us **1.5x** faster.

File: lib/main.js [hunk 1]
```diff
@@ -111,67 +114,29 @@ HTMLUglify.prototype.processRules = function(rules, lookups) {
...
-          var newSelectorPart = part.replace(regex, pointer);
-          var newSelector = rule.selector.replace(part, newSelectorPart);
-          rule.selector = newSelector;
-        }
-      });
+      postcssSelectorParser(function(selectors) {
+        selectors.eachInside(function(selector) {
+          var pointer;
+
+          if ((selector.type === 'id')
```
Comment (elbuo8) [inline lib/main.js:124]: Ouch.

File: package.json [hunk 2]
```diff
@@ -20,10 +20,11 @@
   "dependencies": {
     "cheerio": "^0.19.0",
     "hashids": "^1.0.2",
-    "postcss-safe-parser": "^1.0.1"
+    "postcss-safe-parser": "^1.0.4",
+    "postcss-selector-parser": "^1.3.0"
   },
   "devDependencies": {
-    "benchmark": "^1.0.0",
+    "benchmark": "^2.0.0",
```
Comment (elbuo8) [inline package.json]: Not needed?

**Comment by elbuo8** [comment]
Looking good so far. :dancers:

File: lib/main.js (hunk 1 above)
Comment (devinus) [inline lib/main.js:124]: ?

**Comment by devinus** [comment]
Also microbenchmarking and profiling was done using https://nodejs.org/en/docs/guides/simple-profiling/

I noticed a lot of time was spent in cheerio parsing itself, so I started looking into `posthtml`.

File: lib/main.js (hunk 1 above)
Comment (elbuo8) [inline lib/main.js:124]: Complicated `if` haha.

File: test/main.js [hunk 3]
```diff
@@ -128,11 +128,11 @@ describe('HTMLUglify', function() {
       assert.equal(pointer, 'en');
     });
     it('finds a more complex existing pointer', function() {
-      var lookups = { 
-        class: { 
-          test: 'xz', 
-          testOther: 'wk', 
-          otratest: 'en'
```
Comment (motdotla) [inline test/main.js:135]: thanks for the cleanup :bow:

**Comment by motdotla** [comment]
:+1: nice amount of lines removed

**Pull request merged**